catalyst/contrib/nn/criterion/ntxent.py

```diff
@@ -0,0 +1,61 @@
+from typing import List, Union
+
+import torch
+from torch import nn
+
+
+class NTXentLoss(nn.Module):
+    """
+    NTXent loss from `A Simple Framework for Contrastive Learning of Visual Representations` paper.
+
+    Adapter from:
+    https://arxiv.org/abs/2002.05709
+
+    Example:
+        >>> import torch
+        >>> from torch.nn import functional as F
+        >>> from catalyst.contrib.nn import NTXent
+        >>>
+        >>> features = F.normalize(torch.rand(256, 64, requires_grad=True))
+        >>> labels = torch.randint(high=10, size=(256))
+        >>> criterion = NTXent(tau=0.25)
+        >>> criterion(features1, features2)
+    """
+
+    def __init__(self, tau: float) -> None:
+        """
+
+        Args:
+            tau: tau to use
+        """
+        super().__init__()
+        self.tau = tau
+        self.cosineSim = nn.CosineSimilarity()
+
+    def forward(self, features1: torch.Tensor, features2: torch.Tensor) -> torch.Tensor:
+        """
+
+        Args:
+            features1: batch with samples features of shape
+                [bs; feature_len]
+            features2: batch with samples features of shape
+                [bs; feature_len]
+
+        Returns:
+            torch.Tensor: NTXent loss
+        """
+        assert (
+            features1.shape == features2.shape
+        ), f"Invalid shape of input features: {features1.shape} and {features2.shape}"
+        bs = features1.shape[0]
+
+        pos_loss = self.cosineSim(features1, features2).sum(dim=0) / self.tau
+        list_neg_loss = [
+            torch.exp(self.cosineSim(features1, torch.roll(features2, i, 1)) / self.tau)
+            for i in range(0, bs)
+        ]
+        # todo torch.exp(self.cosineSim(features1, torch.roll(features2, i, 1)))/ self.tau for i in range(0, bs)
+        neg_loss = torch.log(torch.stack(list_neg_loss, dim=0).sum(dim=0)).sum(dim=0)
```

Comment:
Would you kindly explain your opinion?

Reply:
yes, of course. Let me try my math skills.
Let f1 - features1
f2 - features2
n - last index (i.e. f1 and f2 of shape [n+1, feature_size])
<x1, x2> - is exp(cos_sim(x1,x2))
let \tau = 1 for simplicity.
Then, we have, that `list_neg_loss` has next structure:
```
[
 [<f1[0], f2[0]>, ... , <f1[n], f2[n]>],
 [<f1[0], f2[n]>, <f1[1], f2[0]>, ..., <f1[n], f2[n-1]>],
 ...
 [<f1[0], f2[1]>, ..., <f1[n], f2[0]>],
]
```
then, according to this code we have in the denominator:
```
neg_loss = \sum_{k=0}^{n}log\sum_{i=0}^{n}[<f1[i], f2[(i-k)%(n+1)]>]
```
It's not quite right, because, according to the paper (check me if I'm wrong) index of f1 inside log function must be fixed. I.e we should have something like:

```
neg_loss = \sum_{k=0}^{n}log\sum_{i=0}^{n}[<f1[k], f2[i]>]
```

And I think it will be true, if you change `.sum(dim=1)`, except for elements rearangement in the inner summation (please check my statements thoroughly)